### PR TITLE
libvpx: support for conan v2

### PR DESCRIPTION
**Pull request by paulharris**
Started work, got this weird message and I'm wondering if conan is injecting --bindir 

Any advice please?

```
libvpx/1.11.0: Calling:
 > "/build/conandata/conan-data/libvpx/1.11.0/_/_/build/a7d7b9773babc7a104bc397939b61bb2babf96bc/src/configure" '--disable-shared' '--enable-static' '--prefix=/' '--bindir=${prefix}/bin' '--sbindir=${prefix}/bin' '--libdir=${prefix}/lib' '--includedir=${prefix}/include' '--oldincludedir=${prefix}/include' '--disable-examples' '--disable-unit-tests' '--disable-tools' '--disable-docs' '--enable-vp9-highbitdepth' '--as=yasm' '--target=x86_64-linux-gcc' '--disable-avx' '--disable-avx2' '--disable-avx512' 
  disabling shared
Unknown option "--bindir=${prefix}/bin".
```

### CODE REVIEW

**Comment by paulharris** [comment]
I thought I should document this somewhere...
perhaps there is a better way to solve this problem?


libvpx's configure seems to do strange things.  Strange to me anyway.

On Linux, libvpx's "configure" doesn't understand --bindir and --includedir, so I "fixed it" by adding:
```
def layout(self):
        basic_layout(self, src_folder="src")
        self.cpp.package.bindirs = []
        self.cpp.package.includedirs = []
```
and, it doesn't seem to substitute ${prefix}

This results in:
```    
libvpx/1.11.0: Calling:
 > "/build/conandata/conan-data/libvpx/1.11.0/_/_/build/a7d7b9773babc7a104bc397939b61bb2babf96bc/src/configure" '--disable-shared' '--enable-static' '--prefix=/' '--libdir=${prefix}/lib' '--disable-examples' '--disable-unit-tests' '--disable-tools' '--disable-docs' '--enable-vp9-highbitdepth' '--as=yasm' '--target=x86_64-linux-gcc' '--disable-avx' '--disable-avx2' '--disable-avx512' 
...snip...
Libdir ${prefix}/lib must be a subdirectory of 
...snip...
	ConanException: Error 1 while executing "/build/conandata/conan-data/libvpx/1.11.0/_/_/build/a7d7b9773babc7a104bc397939b61bb2babf96bc/src/configure" '--disable-shared' '--enable-static' '--prefix=/' '--libdir=${prefix}/lib' '--disable-examples' '--disable-unit-tests' '--disable-tools' '--disable-docs' '--enable-vp9-highbitdepth' '--as=yasm' '--target=x86_64-linux-gcc' '--disable-avx' '--disable-avx2' '--disable-avx512' 
```

So I add `self.cpp.package.libdirs = []`
and then I add:
```
        tc.configure_args.extend([
            "--prefix=/output",
            "--libdir=/output/lib",
```

It then builds and puts the output files into:
`/build/conandata/conan-data/libvpx/1.11.0/_/_/package/a7d7b9773babc7a104bc397939b61bb2babf96bc/output/include/vpx/vp8.h`
Note the extra **output**.
(I originally tried self.package_folder, and the result was `/packagefolder/packagefolder/lib`)

It would be idea if that wasn't added, but if I add what is effectively the defaults:
```
        tc.configure_args.extend([
            "--prefix=/",
            "--libdir=/lib",
```
Then I get this complaint from configure:
`Libdir /lib must be a subdirectory of`
Which is a result of this bit of code in libvpx's configure:
```
post_process_common_cmdline() {
  prefix="${prefix:-/usr/local}"
  prefix="${prefix%/}"
  libdir="${libdir:-${prefix}/lib}"
  libdir="${libdir%/}"
  if [ "${libdir#${prefix}}" = "${libdir}" ]; then
    die "Libdir ${libdir} must be a subdirectory of ${prefix}"
  fi
}
```

A simple `/` prefix and `/lib` is apparently not enough to satisfy this test.

So my end "solution" is:
```
def generate(self):
        tc.configure_args.extend([
            "--prefix=/output_goes_here",
            "--libdir=/output_goes_here/lib",
...
def package(self):
        rename(self,
                os.path.join(self.package_folder, "output_goes_here", "include"),
                os.path.join(self.package_folder, "include")
                )
        rename(self,
                os.path.join(self.package_folder, "output_goes_here", "lib"),
                os.path.join(self.package_folder, "lib")
                )
        rmdir(self, os.path.join(self.package_folder, "output_goes_here"))
```

I literally used "output_goes_here" for easy grepping.

**Comment by paulharris** [comment]
note that the MSVC conversion hasn't begun yet ...

**Comment by paulharris** [comment]
Recipe is getting close, would like to start getting some reviews please.
Especially on the whole LTO / -GL flag detection in CFLAGS, that is all above my head.
I wonder how to enable LTO anyway ?

**Comment by paulharris** [comment]
@prince-chrismc linter doesn't like the line:
`from conan.tools.microsoft.visual import msvc_version_to_vs_ide_version`
https://github.com/conan-io/conan-center-index/actions/runs/3404784548/jobs/5662271149
`Error: recipes/libvpx/all/conanfile.py:9: [E9011(conan-import-tools), ] Import tools following pattern 'from conan.tools.xxxx import yyyyy' (https://docs.conan.io/en/latest/reference/conanfile/tools.html).`

File: recipes/libvpx/all/conanfile.py
Comment (prince-chrismc) [inline recipes/libvpx/all/conanfile.py:9]: another case where https://github.com/conan-io/conan/issues/12322 would be helpful 

linter complains because it;s private

**Comment by paulharris** [comment]
Any idea why the build is failing?  It can't find pthread symbols,
but pthread was added to self.cpp_info.system_libs()
The problem only appears on clang-linux,...

https://c3i.jfrog.io/c3i/misc/logs/pr/13799/13-linux-clang/libvpx/1.9.0//017b0bf62dc7b32c078b1e271f959f3f897204b0-build.txt

I'm not sure why this fails, it works on Proof-of-conan (with clang-10):
https://github.com/eirikb/proof-of-conan/actions/runs/3584841895/jobs/6032040277

**Comment by paulharris** [comment]
Sorry for the force-push.
This PR now builds locally and on Proof-of-Conan with the problematic compiler/environment.
Without any change.
Not sure why, very confusing.  It failed to build this morning, and now it just started to build correctly.

**Comment by paulharris** [comment]
This is now ready for reviews!

File: recipes/libvpx/all/conandata.yml
Comment (MartinDelille) [inline recipes/libvpx/all/conandata.yml:15]:
```suggestion
      patch_description: "Add support for more compilers"
```

File: recipes/libvpx/all/conandata.yml
Comment (MartinDelille) [inline recipes/libvpx/all/conandata.yml:19]:
```suggestion
      patch_description: "Add support for more compilers"
```

File: recipes/libvpx/all/conandata.yml
Comment (MartinDelille) [inline recipes/libvpx/all/conandata.yml:23]:
```suggestion
      patch_description: "Add support for more compilers"
```

File: recipes/libvpx/all/conandata.yml
Comment (MartinDelille) [inline recipes/libvpx/all/conandata.yml:26]:
```suggestion
      patch_description: "Add support for more compilers"
```

File: recipes/libvpx/all/conanfile.py
Comment (MartinDelille) [inline recipes/libvpx/all/conanfile.py:231]: A similar issue has been raised here by @sorny92 : https://github.com/conan-io/conan/issues/12431

**Review by MartinDelille — CHANGES_REQUESTED** [review]

File: recipes/libvpx/all/conanfile.py
Comment (paulharris) [reply recipes/libvpx/all/conanfile.py:231]: Yes there is a few issues open on that theme

Comment (paulharris) [reply recipes/libvpx/all/conanfile.py:231]: My plan is to finish this PR, and someone can cycle back and fix it up with the new stuff when it has landed.
I know Autotools probably has some work that needs to be done in Jan anyway.

Comment (MartinDelille) [reply recipes/libvpx/all/conanfile.py:231]: Nice job by the way! I had a bit of struggle trying to make a conan v2 version of Ffmpeg and your work may be very helpful!

**Review by MartinDelille — DISMISSED** [review]
I'm just a community reviewer and it looks good to me but let's see what CCI maintainers have to say about it!

**Review by prince-chrismc — APPROVED** [review]
There's a lot of work here, I did not see any blockers so i'd suggest leaving any smaller issue for the next PR :)

**Comment by prince-chrismc** [comment]
The linter check is required but blocked since its a private import

**Comment by MartinDelille** [comment]
If I understand correctly, this recipe could be use `AutotoolsToolchain` once we will be using conan 1.57.0, right ? 

- https://github.com/conan-io/conan/releases/tag/1.57.0
- https://github.com/conan-io/conan/pull/12884

**Comment by paulharris** [comment]
> If I understand correctly, this recipe could be use `AutotoolsToolchain` once we will be using conan 1.57.0, right ?
> 
> * https://github.com/conan-io/conan/releases/tag/1.57.0
> * [[AutotoolsToolchain] Added helpers `update_xxxx_args` conan#12884](https://github.com/conan-io/conan/pull/12884)

Do you mean, for building for Windows?  I'm not sure, it would be nice to just use one build method...

**Comment by paulharris** [comment]
> The linter check is required but blocked since its a private import

So, how to progress this?  Duplicate the logic?

**Comment by SSE4** [comment]
> > The linter check is required but blocked since its a private import
> 
> So, how to progress this? Duplicate the logic?

create an issue in conan repo to expose `msvc_version_to_vs_ide_version`. if it's needed for recipes, it's needed. duplicating logic everywhere is not an option.
for the time being, I think linter can be suppressed temporarily (@prince-chrismc there were some sort of `no-lint` directive?)

**Comment by SSE4** [comment]
> AutotoolsToolchain

AFAIK `libvpx` is not autotools-based. it uses its own scripts, "similar" to autotools, but written from scratch. it understands some common arguments and variables, but not all of them, unfortunately.

**Comment by prince-chrismc** [comment]
I raised the issue to the team but missed the last sync so I am not sure the status. I'll ping them again if there's no response

**Comment by paulharris** [comment]
> > AutotoolsToolchain
> 
> AFAIK `libvpx` is not autotools-based. it uses its own scripts, "similar" to autotools, but written from scratch. it understands some common arguments and variables, but not all of them, unfortunately.

This is correct.  So we have to use their MS Build stuff for Windows.

**Comment by prince-chrismc** [comment]
added to weekly meeting for the team to review :)

File: recipes/libvpx/all/conanfile.py
Comment (prince-chrismc) [reply recipes/libvpx/all/conanfile.py:9]: image

For sheppards

Comment (prince-chrismc) [reply recipes/libvpx/all/conanfile.py:9]:
```py
if is_msvc(self) and Version(self.settings.compiler.version) > "100":
    ide_version = {"190": "14", "191": "15", "192": "16"}
```

Comment (jcar87) [reply recipes/libvpx/all/conanfile.py:9]: As @prince-chrismc  currently the method is private and the linter is rightfully warning us about this. This is a good case to make it public, but if that were to be the case it would end up with a different import pattern.

@prince-chrismc 's suggestion above (or something similar) might be enough to unblock this. If it works satisfactorily, please mention the issue in a comment next to the code so that we can clean this up in the future! Thanks :)

**Comment by paulharris** [comment]
Thanks @prince-chrismc for following up with this, have been busy elsewhere so I'm very happy to see progress!

**Comment by prince-chrismc** [comment]
I totally feel you, super glad to get this over the finish

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
Tricky recipe, thank you for detailing all questions!

File: recipes/libvpx/all/conanfile.py
Comment (SpaceIm) [inline recipes/libvpx/all/conanfile.py:73]: this trick defeats `fix_apple_shared_install_name()` and even the hook checking whether shared libs are relocatables.

Comment (paulharris) [reply recipes/libvpx/all/conanfile.py:73]: I might be able to use `tc.update_configure_args({"--bindir":None})` instead (or something like that).
I won't be able to look for another ~12 hours, if you have the time please feel free to fix up and suggest more :)

Comment (paulharris) [reply recipes/libvpx/all/conanfile.py:73]: Actually, won't have time to fix, doing the other recipes.  Try my suggestion?  Happy to test your PR locally, tag me in it.

**Pull request merged**
